File: 2nd/homework_3/id_69/114_FlattenBinaryTreeToLinkedList.py
# author: lu19kend
# https://leetcode.com/problems/flatten-binary-tree-to-linked-list/description/


class Solution(object):
    prev = None

    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if root is None:
            return
        self.flatten(root.right)
        self.flatten(root.left)
        root.right = self.prev
        root.left = None
        self.prev = root

File: 2nd/homework_3/id_45/114_flatten_binary_tree_to_linked_list.cpp
/**
 *  * Definition for a binary tree node.
 *  * struct TreeNode {
 *  *     int val;
 *  *     TreeNode *left;
 *  *     TreeNode *right;
 *  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *  * };
 *  */
//right link
class Solution {
public:
    void flatten(TreeNode* root) {
        while (root) {
            if (root->left && root->right) {
                TreeNode *t = root->left;
                while(t->right)
                    t = t->right;
                t->right = root->right;
            }

            if (root->left) {
                root->right = root->left;
                root->left = NULL;
            }
            root = root->right;
        }   
    }
};
